File: srcs/server/agents/builtin/toolsearch.go
package builtin

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
)

// toolManifest is a static manifest used by ToolSearch.
// It mirrors AllTools + coordinator tools without creating an init cycle.
var toolManifest = []struct{ Name, Desc, Params string }{
	{"Bash", "Execute a bash command. Use for build/test/git/shell operations.", ""},
	{"Read", "Read a file from the local filesystem.", ""},
	{"Write", "Create or overwrite a file on the local filesystem.", ""},
	{"Edit", "Edit a file: replace an exact string with a new string.", ""},
	{"Glob", "List files matching a glob pattern.", ""},
	{"Grep", "Search for a regex pattern in files.", ""},
	{"WebFetch", "Fetch a URL and return its content as text.", ""},
	{"WebSearch", "Search the web using DuckDuckGo.", ""},
	{"SendMessage", "Send a message to the user or another agent.", ""},
	{"TodoWrite", "Create or update the session todo list.", ""},
	{"TodoRead", "Read the current session todo list.", ""},
	{"ToolSearch", "Search available tools by name or description.", ""},
	{"TaskCreate", "Create a new task in the task list.", ""},
	{"TaskGet", "Get details of a task by ID.", ""},
	{"TaskList", "List all tasks.", ""},
	{"TaskUpdate", "Update a task (subject, description, status, owner, etc.).", ""},
	{"Sleep", "Sleep for N seconds; use while waiting for CI/builds.", ""},
	{"Agent", "Spawn a background sub-agent to perform a task concurrently.", ""},
	{"TaskStop", "Stop a running background agent task.", ""},
	{"TaskStatus", "Get the status and progress of a background agent task.", ""},
}

func getAllToolsManifest() []struct{ Name, Description, Parameters string } {
	var res []struct{ Name, Description, Parameters string }
	for _, m := range toolManifest {
		res = append(res, struct{ Name, Description, Parameters string }{m.Name, m.Desc, m.Params})
	}
	return res
}

// ToolSearchTool definition - returns available tools matching an optional query.
// Mirrors CC-Source's ToolSearchTool which searches available tool definitions.
var ToolSearchTool = Tool{
	Name: "ToolSearch",
	Description: "Search for available tools and their descriptions. " +
		"Use this when you are unsure which tool to use for a given task.",
	Parameters: json.RawMessage(`{
		"type": "object",
		"properties": {
			"query": {
				"type": "string",
				"description": "Optional search query to filter tools by name or description."
			}
		}
	}`),
	Execute: func(ctx context.Context, args json.RawMessage) (string, error) {
		var input struct {
			Query string `json:"query"`
		}
		if len(args) > 0 {
			if err := json.Unmarshal(args, &input); err != nil {
				return "", err
			}
		}

		q := strings.ToLower(input.Query)

		isSelect := strings.HasPrefix(q, "select:")
		var requiredNames []string
		if isSelect {
			parts := strings.Split(strings.TrimPrefix(q, "select:"), ",")
			for _, p := range parts {
				p = strings.TrimSpace(p)
				if p != "" {
					requiredNames = append(requiredNames, strings.ToLower(p))
				}
			}
		}

		var results []string
		manifest := getAllToolsManifest()
		seen := make(map[string]bool)
		for _, t := range manifest {
			if seen[t.Name] {
				continue
			}
			seen[t.Name] = true

			match := false
			if isSelect {
				for _, req := range requiredNames {
					if strings.ToLower(t.Name) == req {
						match = true
						break
					}
				}
			} else {
				if q == "" ||
					strings.Contains(strings.ToLower(t.Name), q) ||
					strings.Contains(strings.ToLower(t.Description), q) {
					match = true
				}
			}

			if match {
				desc := t.Description
				schema := string(t.Parameters)
				if schema == "" {
					schema = "{}"
				}
				results = append(results, fmt.Sprintf("<function>{\"name\": %q, \"description\": %q, \"parameters\": %s}</function>", t.Name, desc, schema))
			}
		}

		if len(results) == 0 {
			return "No tools found matching query.", nil
		}

		return fmt.Sprintf("<functions>\n%s\n</functions>", strings.Join(results, "\n")), nil
	},
}

File: src/agents/builtin/tools/toolsearch.rs
use ohc_builtin_agent_core::types::ToolError;
use serde_json::json;
use serde::Deserialize;
use std::sync::Arc;

use super::{Tool, pydantic::{PydanticToolExecutor, PydanticAdapter}};

#[derive(Deserialize)]
struct ToolSearchArgs {
    query: String,
}


struct ToolSearchExecutor;

#[async_trait::async_trait]
impl PydanticToolExecutor<ToolSearchArgs> for ToolSearchExecutor {
    async fn execute_typed(&self, args: ToolSearchArgs) -> Result<String, ToolError> {
        let query = args.query.to_lowercase();

        let all_tools = &[
            ("Bash", "Execute shell commands"),
            ("Read", "Read file contents"),
            ("Write", "Write content to a file"),
            ("Edit", "Edit a file by replacing text"),
            ("Glob", "Find files by pattern"),
            ("Grep", "Search file contents with regex"),
            ("WebFetch", "Fetch URL content"),
            ("WebSearch", "Search the web"),
            ("SendMessage", "Send message to parent agent"),
            ("TodoWrite", "Write the task todo list"),
            ("TodoRead", "Read the task todo list"),
            ("ToolSearch", "Search available tools"),
            ("TaskCreate", "Create a new task"),
            ("TaskGet", "Get task details"),
            ("TaskList", "List all tasks"),
            ("TaskUpdate", "Update task status/result"),
            ("Sleep", "Sleep for N seconds"),
            ("Agent", "Spawn a sub-agent task"),
            ("TaskStop", "Stop a running sub-agent"),
            ("TaskStatus", "Get sub-agent task status"),
            (
                "RunWorkflow",
                "Run a built-in multi-agent workflow with phased subagent orchestration",
            ),
        ];

        let matches: Vec<String> = all_tools
            .iter()
            .filter(|(name, desc)| {
                name.to_lowercase().contains(&query) || desc.to_lowercase().contains(&query)
            })
            .map(|(name, desc)| format!("{}: {}", name, desc))
            .collect();

        if matches.is_empty() {
            Ok(format!("No tools found matching '{}'.", query))
        } else {
            Ok(matches.join("\n"))
        }
    }
}

pub fn toolsearch_tool() -> Tool {
    Tool {
        name: "ToolSearch".to_string(),
        description: "Search available tools by name or description.".to_string(),
        is_read_only: true,
        parameters: json!({
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Search term to find relevant tools."
                }
            },
            "required": ["query"]
        }),
        execute: Arc::new(PydanticAdapter::new(ToolSearchExecutor)),
    }
}
